File: advanced-algorithms/arrays/prefix-sums/find_pivot_index.cpp
#include <vector>

using namespace std;

/*
 * Author: Valerie Doan
 * Date: October 16, 2024
 * Problem: Find Pivot Index
 * Level: Easy
 * Neetcode Course: Advanced Algorithms
 * Neetcode Topic: Prefix Sums
 */

int pivotIndex(vector<int> &nums)
{
    vector<int> prefix;
    int total = 0;

    for (int num : nums)
    {
        total += num;
        prefix.push_back(total);
    }

    vector<int> postfix;
    total = 0;

    for (int i = nums.size() - 1; i >= 0; i--)
    {
        total += nums[i];
        postfix.push_back(total);
    }

    for (int i = 0; i < nums.size(); i++)
    {
        if (prefix[i] == postfix[nums.size() - 1 - i])
        {
            return i;
        }
    }

    return -1;
}

File: python/advanced-algorithms/arrays/prefix-sums/find_pivot_index.py
"""
Author: Valerie Doan
Date: November 7, 2024
Problem: Find Pivot Index
Level: Easy
Neetcode Course: Advanced Algorithms
Neetcode Topic: Prefix Sum
"""


class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        prefix = [0] * len(nums)
        postfix = [0] * len(nums)

        for i in range(len(nums)):
            if i == 0:
                prefix[i] = nums[i]
            else:
                prefix[i] = prefix[i - 1] + nums[i]

        for i in range(len(nums) - 1, -1, -1):
            if i == len(nums) - 1:
                postfix[i] = nums[i]
            else:
                postfix[i] = postfix[i + 1] + nums[i]

        for i in range(len(nums)):
            if prefix[i] == postfix[i]:
                return i

        return -1
